applications/FluidDynamicsApplication/custom_processes/embedded_nodes_initialization_process.h

```diff
@@ -0,0 +1,334 @@
+                {
+                    NodeType::Pointer pNode;
+                    double p_avg = 0.0;
+                    array_1d<double, 3> v_avg;
+                    v_avg[0] = 0.0;
+                    v_avg[1] = 0.0;
+                    v_avg[2] = 0.0;
+
+                    for (unsigned int j=0; j<ElemNumNodes; ++j)
+                    {
+                        if (rGeometry[j].IsNot(SELECTED))
+                        {
+                            // Compute the average velocity in the non selected nodes
+                            p_avg += rGeometry[j].FastGetSolutionStepValue(PRESSURE);
+                            v_avg += rGeometry[j].FastGetSolutionStepValue(VELOCITY);
+                        }
+                        else
+                        {
+                            // Get a pointer to the unique SELECTED node
+                            NodeType::Pointer pAux = rGeometry(j);
+                            std::swap(pAux, pNode);
+                        }
+                    }
+
+                    // Compute the non-SELECTED nodes average values
+                    p_avg /= (ElemNumNodes-1);
+                    v_avg /= (ElemNumNodes-1);
+
+                    // Historical values initialization
+                    pNode->SetLock();                                               // Avoid another thread to acces the node while the value set
+                    pNode->Set(SELECTED, false);                                    // Once a node has been initialized it is marked as non SELECTED
+                    for (unsigned int step=0; step<BufferSize; ++step)              // Fill the velocity and pressure buffer
+                    {
+                        pNode->FastGetSolutionStepValue(PRESSURE, step) = p_avg;
+                        pNode->FastGetSolutionStepValue(VELOCITY, step) = v_avg;
+
+                    }
+                    pNode->UnSetLock();                                             // Free the node thread access
+                }
+            }
+        }
+
+        array_1d<double, 3> v_null;
+        v_null[0] = 0.0;
+        v_null[1] = 0.0;
+        v_null[2] = 0.0;
+
+        // If there still exist some remaining nodes, initialize their values to zero
+        #pragma omp parallel for firstprivate(v_null)
+        for (int k = 0; k < static_cast<int>(rNodes.size()); ++k)
+        {
+            ModelPart::NodesContainerType::iterator itNode = rNodes.begin() + k;
+
+            if (itNode->Is(SELECTED))
+            {
+                itNode->Set(SELECTED, false);                                   // Once a node has been initialized it is marked as non SELECTED
+                for (unsigned int step=0; step<BufferSize; ++step)              // Fill the velocity and pressure buffer
+                {
+                    itNode->FastGetSolutionStepValue(PRESSURE, step) = 0.0;
+                    itNode->FastGetSolutionStepValue(VELOCITY, step) = v_null;
```

Comment:
here maybe it is even better to do

          noalias(itNode->FastGetSolutionStepValue(VELOCITY, step)) = ZeroVector(3)